File: redwood-yoga-java/src/main/java/io/github/orioncraftmc/meditate/internal/enums/YGExperiment.java
package io.github.orioncraftmc.meditate.internal.enums;

import java.util.EnumSet;
import java.util.Set;

public enum YGExperiment {

    kDoubleMeasureCallbacks;

    private static final Set<YGExperiment> ENABLED_EXPERIMENTS = EnumSet.noneOf(YGExperiment.class);

    public static void disableAllExperiments() {
        ENABLED_EXPERIMENTS.clear();
    }

    public void enable() {
        ENABLED_EXPERIMENTS.add(this);
    }

    public void disable() {
        ENABLED_EXPERIMENTS.remove(this);
    }

    public boolean toggle() {
        if (ENABLED_EXPERIMENTS.contains(this)) {
            return ENABLED_EXPERIMENTS.remove(this);
        } else {
            ENABLED_EXPERIMENTS.add(this);
            return false;
        }
    }

    public boolean isEnabled() {
        return ENABLED_EXPERIMENTS.contains(this);
    }
}

File: redwood-yoga/src/commonMain/kotlin/app/cash/redwood/yoga/enums/YGExperiment.kt
package app.cash.redwood.yoga.enums

enum class YGExperiment {
  kDoubleMeasureCallbacks;

  fun enable() {
    ENABLED_EXPERIMENTS.add(this)
  }

  fun disable() {
    ENABLED_EXPERIMENTS.remove(this)
  }

  fun toggle(): Boolean {
    return if (ENABLED_EXPERIMENTS.contains(this)) {
      ENABLED_EXPERIMENTS.remove(this)
    } else {
      ENABLED_EXPERIMENTS.add(this)
      false
    }
  }

  fun isEnabled(): Boolean {
    return ENABLED_EXPERIMENTS.contains(this)
  }

  companion object {
    private val ENABLED_EXPERIMENTS: MutableSet<YGExperiment> = mutableSetOf()

    fun disableAllExperiments() {
      ENABLED_EXPERIMENTS.clear()
    }
  }
}
